web/html/src/components/hub/RegisterPeripheralForm.tsx

```diff
@@ -0,0 +1,276 @@
+import * as React from "react";
+
+import { AsyncButton, SubmitButton } from "components/buttons";
+import { Form, Password, Radio, Text, TextArea } from "components/input";
+import { Messages, MessageType } from "components/messages/messages";
+import { TopPanel } from "components/panels";
+import Validation from "components/validation";
+
+import { Utils } from "utils/functions";
+import Network from "utils/network";
+
+import { HubRegisterRequest } from "./types";
+
+enum RegistrationMode {
+  Token = "token",
+  UserPassword = "userPassword",
+}
+
+enum CertificateMode {
+  NotNeeded = "notNeeded",
+  Upload = "upload",
+  Paste = "paste",
+}
+
+type FormModel = {
+  serverFqdn: string;
+  registrationMode: RegistrationMode;
+  token?: string;
+  username?: string;
+  password?: string;
+  certificateMode: CertificateMode;
+  uploadedRootCA?: string;
+  pastedRootCA?: string;
+};
+
+type Props = {};
+
+type State = {
+  model: FormModel;
+  messages: MessageType[];
+  loading: boolean;
+  validated: boolean;
+};
+
+export class RegisterPeripheralForm extends React.Component<Props, State> {
+  // Initial form model
+  private static readonly INITIAL_MODEL: FormModel = {
+    serverFqdn: "",
+    registrationMode: RegistrationMode.Token,
+    certificateMode: CertificateMode.NotNeeded,
+  };
+
+  constructor(props: Props) {
+    super(props);
+
+    this.state = {
+      model: RegisterPeripheralForm.INITIAL_MODEL,
```

Comment:
We're planning to refactor some form related code for 5.1, it will be easier for us to refactor with the below change instead, but you could not have known that. 🙂

```suggestion
      model: { ...RegisterPeripheralForm.INITIAL_MODEL },
```

Reply:
fixed!